File: panel/models/reactive_html.py
import difflib
import re

from collections import defaultdict
from html.parser import HTMLParser

import bokeh.core.properties as bp

from bokeh.events import ModelEvent
from bokeh.model import DataModel
from bokeh.models import LayoutDOM

from .layout import HTMLBox

endfor = '{%-? endfor -?%}'
list_iter_re = r'{%-? for (\s*[A-Za-z_]\w*\s*) in (\s*[A-Za-z_]\w*\s*) -?%}'
items_iter_re = r'{%-? for \s*[A-Za-z_]\w*\s*, (\s*[A-Za-z_]\w*\s*) in (\s*[A-Za-z_]\w*\s*)\.items\(\) -?%}'
values_iter_re = r'{%-? for (\s*[A-Za-z_]\w*\s*) in (\s*[A-Za-z_]\w*\s*)\.values\(\) -?%}'


class ReactiveHTMLParser(HTMLParser):

    def __init__(self, cls, template=True):
        super().__init__()
        self.template = template
        self.cls = cls
        self.attrs = defaultdict(list)
        self.children = {}
        self.nodes = []
        self.looped = []
        self._template_re = re.compile(r'\$\{[^}]+\}')
        self._literal_re = re.compile(r'\{\{[^}]+\}\}')
        self._current_node = None
        self._node_stack = []
        self._open_for = False
        self.loop_map = {}
        self.loop_var_map = defaultdict(list)

    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        dom_id = attrs.pop('id', None)
        self._current_node = None
        self._node_stack.append((tag, dom_id))

        if not dom_id:
            for attr, value in attrs.items():
                if value is None:
                    continue
                params, methods = [], []
                for match in self._template_re.findall(value):
                    match = match[2:-1]
                    if match.startswith('model.'):
                        continue
                    if match in self.cls.param:
                        params.append(match)
                    elif hasattr(self.cls, match):
                        methods.append(match)
                if methods:
                    raise ValueError(
                        "DOM nodes with an attached callback must declare "
                        f"an id. Found <{tag}> node with the `{attr}` callback "
                        f"referencing the `{methods[0]}` method. Add an id "
                        "attribute like this: "
                        f"<{tag} id=\"{tag}\" {attr}=\"${{{methods[0]}}}>...</{tag}>."
                    )
                elif params:
                    literal = value.replace(f'${{{params[0]}}}', f'{{{{{params[0]}}}}}')
                    raise ValueError(
                        "DOM node with a linked parameter declaration "
                        f"must declare an id. Found <{tag}> node with "
                        f"the `{attr}` attribute referencing the `{params[0]}` "
                        "parameter. Either declare an id on the node, "
                        f"i.e. <{tag} id=\"{tag}\" {attr}=\"{value}\">...</{tag}>, "
                        "or insert the value as a literal: "
                        f"<{tag} {attr}=\"{literal}\">...</{tag}>."
                    )
            return

        if dom_id in self.nodes:
            raise ValueError(f'Multiple DOM nodes with id="{dom_id}" found.')
        self._current_node = dom_id
        self.nodes.append(dom_id)
        for attr, value in attrs.items():
            if value is None:
                continue
            matches = []
            for match in self._template_re.findall(value):
                if not match[2:-1].startswith('model.'):
                    matches.append(match[2:-1])
            if matches:
                self.attrs[dom_id].append((attr, matches, value.replace('${', '{')))

    def handle_endtag(self, tag):
        self._node_stack.pop()
        self._current_node = self._node_stack[-1][1] if self._node_stack else None

    def handle_data(self, data):
        if not self.template:
            return

        dom_id = self._current_node
        matches = []
        for match in self._template_re.findall(data):
            var = match[2:-1].strip()
            if match[2:-1] not in self.loop_var_map[var]:
                self.loop_var_map[var].append(match[2:-1])
            if var.endswith('.index0'):
                matches.append('${%s }}]}' % var)  # noqa: UP031
            else:
                matches.append('${%s}' % var)  # noqa: UP031

        literal_matches = []
        for match in self._literal_re.findall(data):
            match = match[2:-2].strip()
            if match.endswith('.index0'):
                literal_matches.append('{{%s }}]}' % match)  # noqa: UP031
            else:
                literal_matches.append('{{ %s }}' % match)  # noqa: UP031

        # Detect templating for loops
        list_loop = re.findall(list_iter_re, data)
        values_loop = re.findall(values_iter_re, data)
        items_loop = re.findall(items_iter_re, data)
        nloops = len(list_loop) + len(values_loop) + len(items_loop)
        if nloops > 1 and nloops and self._open_for:
            raise ValueError('Nested for loops currently not supported in templates.')
        elif nloops:
            loop = [loop for loop in (list_loop, values_loop, items_loop) if loop][0]
            var, obj = loop[0]
            if var in self.cls.param:
                raise ValueError(
                    f'Loop variable {var} clashes with parameter name. '
                    'Ensure loop variables have a unique name. Relevant '
                    f'template section:\n\n{data}'
                )
            self.loop_map[var] = obj

        open_for = re.search(r'{%-? for', data)
        end_for = re.search(endfor, data)
        if open_for:
            if self._current_node is None:
                node = self._node_stack[-1][0]
                raise ValueError(
                    'Loops may only be used inside a DOM node with an assigned ID. '
                    f'The following loop could not be expanded because the <{node}> node '
                    f'did not have an assigned id:\n\n    {data.strip()}'
                )
            self._open_for = True
        if end_for and (not nloops or end_for.start() > open_for.start()):
            self._open_for = False

        if self._current_node and literal_matches:
            if len(literal_matches) == 1:
                literal_match = literal_matches[0][2:-2].strip()
            else:
                literal_match = None

            if literal_match and (literal_match in self.loop_map) and self._open_for:
                literal_match = self.loop_map[literal_match]
                self.looped.append((dom_id, literal_match))

        if not (self._current_node and matches):
            return

        if len(matches) == 1:
            match = matches[0][2:-1].strip()
        else:
            for match in matches:
                mode = self.cls._child_config.get(match, 'model')
                if mode != 'template':
                    raise ValueError(f"Cannot match multiple variables in '{mode}' mode.")
            match = None

        # Handle looped variables
        if match and (match.strip() in self.loop_map or '[' in match) and self._open_for:
            if match.strip() in self.loop_map:
                loop_match = self.loop_map[match.strip()]
                matches[matches.index(f'${{{match}}}')] = f'${{{loop_match}}}'
                match = loop_match
            elif '[' in match:
                match, _ = match.split('[')
            dom_id = dom_id.replace('-{{ loop.index0 }}', '')
            self.looped.append((dom_id, match))

        mode = self.cls._child_config.get(match, 'model')
        if match in self.cls.param and mode != 'template':
            self.children[dom_id] = match
            return

        templates = []
        for match in matches:
            match = match[2:-1]
            if match.startswith('model.'):
                continue
            if match not in self.cls.param and '.' not in match:
                params = difflib.get_close_matches(match, list(self.cls.param))
                raise ValueError(f"{self.cls.__name__} HTML template references "
                                 f"unknown parameter '{match}', similar parameters "
                                 f"include {params}.")
            templates.append(match)
        self.attrs[dom_id].append(('children', templates, data.replace('${', '{')))



def find_attrs(html):
    p = ReactiveHTMLParser()
    p.feed(html)
    return p.attrs


class DOMEvent(ModelEvent):

    event_name = 'dom_event'

    def __init__(self, model, node=None, data=None):
        self.data = data
        self.node = node
        super().__init__(model=model)


class ReactiveHTML(HTMLBox):

    attrs = bp.Dict(bp.String, bp.List(bp.Tuple(bp.String, bp.List(bp.String), bp.String)))

    callbacks = bp.Dict(bp.String, bp.List(bp.Tuple(bp.String, bp.String)))

    children = bp.Dict(bp.String, bp.Either(bp.List(bp.Either(bp.Instance(LayoutDOM), bp.String)), bp.String))

    data = bp.Instance(DataModel)

    events = bp.Dict(bp.String, bp.Dict(bp.String, bp.Bool))

    event_params = bp.List(bp.String)

    html = bp.String()

    looped = bp.List(bp.String)

    nodes = bp.List(bp.String)

    scripts = bp.Dict(bp.String, bp.List(bp.String))

    def __init__(self, **props):
        if 'attrs' not in props and 'html' in props:
            props['attrs'] = find_attrs(props['html'])
        super().__init__(**props)

File: panel/models/reactive_html.ts
import {render} from "preact"
import {useCallback} from "preact/hooks"
import {html} from "htm/preact"

import {div} from "@bokehjs/core/dom"
import type {Dict, Attrs} from "@bokehjs/core/types"
import {assert, unreachable} from "@bokehjs/core/util/assert"
import {enumerate} from "@bokehjs/core/util/iterator"
import {isArray, isString} from "@bokehjs/core/util/types"
import {dict, keys, entries} from "@bokehjs/core/util/object"
import type * as p from "@bokehjs/core/properties"
import {Model} from "@bokehjs/model"
import {UIElement} from "@bokehjs/models/ui/ui_element"

import {dict_to_records} from "./data"
import {serializeEvent} from "./event-to-object"
import {DOMEvent, html_decode} from "./html"
import {HTMLBox, HTMLBoxView} from "./layout"
import {convertUndefined} from "./util"

function serialize_attrs(attrs: Attrs): Attrs {
  const serialized: Attrs = {}
  for (const [attr, value] of entries(attrs)) {
    const serialized_value = (() => {
      if (isString(value)) {
        if (value !== "" && (value === "NaN" || !isNaN(Number(value)))) {
          return Number(value)
        } else if (value === "false" || value === "true") {
          return value === "true" ? true : false
        }
      }
      return value
    })()
    serialized[attr] = serialized_value
  }
  return serialized
}

function escape_regex(string: string) {
  return string.replace(/[-\/\\^$*+?.()|[\]]/g, "\\$&")
}

function extract_token(template: string, str: string, tokens: string[]): {[key: string]: string} | null {
  const token_mapping: Map<string, string> = new Map()
  for (const match of tokens) {
    token_mapping.set(`{${match}}`, "(.*)")
  }

  const token_list = []
  let regexp_template = `^${escape_regex(template)}$`

  // Find the order of the tokens
  let i, token_index, token_entry
  for (const [m, replacement] of token_mapping) {
    token_index = template.indexOf(m)

    // Token found
    if (token_index > -1) {
      regexp_template = regexp_template.replace(m, replacement)
      token_entry = {
        index: token_index,
        token: m,
      }

      for (i = 0; i < token_list.length && token_list[i].index < token_index; i++) {}

      // Insert it at index i
      if (i < token_list.length) {
        token_list.splice(i, 0, token_entry)
      } else {
        token_list.push(token_entry)
      }
    }
  }

  regexp_template = regexp_template.replace(/\{[^{}]+\}/g, ".*")
  const match = new RegExp(regexp_template).exec(str)

  if (match != null) {
    const result: {[key: string]: string} = {}
    // Find your token entry
    for (i = 0; i < token_list.length; i++) {
      result[token_list[i].token.slice(1, -1)] = match[i + 1]
    }
    return result
  } else {
    return null
  }
}

function element_lookup(root: ShadowRoot, el_id: string): HTMLElement | null {
  let el = root.getElementById(el_id)
  if (el == null) {
    el = document.getElementById(el_id)
  }
  return el
}

export class ReactiveHTMLView extends HTMLBoxView {
  declare model: ReactiveHTML

  html: string
  container: HTMLDivElement
  protected _changing: boolean = false
  protected readonly _event_listeners: Map<string, Map<string, (event: Event) => void>> = new Map()
  protected _mutation_observers: MutationObserver[] = []
  protected _script_fns: Map<string, Function> = new Map()
  protected _state: any = {}

  override initialize(): void {
    super.initialize()
    this.html = html_decode(this.model.html) ?? this.model.html
  }

  _recursive_connect(model: Model, update_children: boolean, path: string): void {
    for (const prop of model) {
      const subpath = path.length != 0 ? `${path}.${prop.attr}` : prop.attr
      const obj = prop.get_value()
      if (obj == null) {
        continue
      }
      if (obj instanceof Model) {
        this._recursive_connect(obj, true, subpath)
      }
      this.on_change(prop, () => {
        if (update_children) {
          for (const [node, attr] of entries(this.model.children)) {
            if (attr == prop.attr) {
              let children = prop.get_value() as string | (string | UIElement)[]
              if (!isArray(children)) {
                children = [children]
              }
              this._render_node(node, children)
              return
            }
          }
        }
        if (!this._changing) {
          this._update(subpath)
        }
      })
    }
  }

  override connect_signals(): void {
    super.connect_signals()

    const {children, events} = this.model.properties
    this.on_change(children, async () => {
      this.html = html_decode(this.model.html) ?? this.model.html
      await this.build_child_views()
      this.invalidate_render()
    })
    this._recursive_connect(this.model.data, true, "")
    this.on_change(events, () => {
      this._remove_event_listeners()
      this._setup_event_listeners()
    })
    this.connect_scripts()
  }

  connect_scripts(): void {
    const {id} = this.model.data
    for (const [prop, scripts] of entries(this.model.scripts)) {
      let data_model = this.model.data
      let attr: string
      if (prop.includes(".")) {
        const path = prop.split(".")
        attr = path[path.length - 1]
        for (const p of path.slice(0, -1)) {
          const value = data_model.property(p).get_value()
          assert(value instanceof Model)
          data_model = value
        }
      } else {
        attr = prop
      }
      for (const script of scripts) {
        const decoded_script = html_decode(script) ?? script
        const script_fn = this._render_script(decoded_script, id)
        this._script_fns.set(prop, script_fn)
        if (!(attr in data_model.properties)) {
          continue
        }
        const property = data_model.property(attr)
        const is_event_param = this.model.event_params.includes(prop)
        this.on_change(property, () => {
          if (!this._changing && !(is_event_param && !data_model.property(prop).get_value())) {
            this.run_script(prop)
            if (is_event_param) {
              data_model.setv({[prop]: false})
            }
          }
        })
      }
    }
  }

  run_script(property: string, silent: boolean = false): void {
    const script_fn = this._script_fns.get(property)
    if (script_fn === undefined) {
      if (!silent) {
        console.warn(`Script '${property}' could not be found.`)
      }
      return
    }
    const this_obj: any = {
      get_records(property: string, index: boolean): unknown[] {
        return this.get_records(property, index)
      },
    }
    for (const name of this._script_fns.keys()) {
      this_obj[name] = () => this.run_script(name)
    }
    return script_fn(
      this.model,
      this.model.data,
      this._state,
      this,
      (s: string) => this.run_script(s),
      this_obj,
    )
  }

  get_records(property_name: string, index: boolean = true): unknown[] {
    return dict_to_records(this.model.data.property(property_name), index)
  }

  override disconnect_signals(): void {
    super.disconnect_signals()
    this._remove_event_listeners()
    this._remove_mutation_observers()
  }

  override remove(): void {
    this.run_script("remove", true)
    super.remove()
  }

  override get child_models(): UIElement[] {
    const models = []
    for (const [_parent, children] of entries(this.model.children)) {
      for (const model of children) {
        if (!isString(model)) {
          models.push(model)
        }
      }
    }
    return models
  }

  override _after_layout(): void {
    this.run_script("after_layout", true)
  }

  override render(): void {
    this.empty()
    this._update_stylesheets()
    this._update_css_classes()
    this._apply_styles()
    this._apply_visible()

    this.container = div({style: {display: "contents"}})
    this.shadow_el.append(this.container)
    this._update()
    this._render_children()
    this._setup_mutation_observers()
    this._setup_event_listeners()
    this.run_script("render", true)
  }

  private _send_event(elname: string, attr: string, event: any) {
    const serialized = convertUndefined(serializeEvent(event))
    serialized.type = attr
    this.model.trigger_event(new DOMEvent(elname, serialized))
  }

  private _render_child(child: string | UIElement, el: Element): void {
    if (isString(child)) {
      el.innerHTML = html_decode(child) ?? child
    } else {
      const view = this._child_views.get(child)
      if (view == null) {
        el.innerHTML = ""
      } else {
        el.appendChild(view.el)
        view.render()
        view.r_after_render()
      }
    }
  }

  _render_node(node: string, children: (string | UIElement)[]): void {
    const {id} = this.model.data
    if (this.model.looped.includes(node)) {
      for (const [child, i] of enumerate(children)) {
        const el_id = `${node}-${i}-${id}`
        const el = element_lookup(this.shadow_el, el_id)
        if (el == null) {
          console.warn(`DOM node '${el_id}' could not be found. Cannot render children.`)
          continue
        }
        this._render_child(child, el)
      }
    } else {
      const el_id = `${node}-${id}`
      const el = element_lookup(this.shadow_el, el_id)
      if (el == null) {
        console.warn(`DOM node '${el_id}' could not be found. Cannot render children.`)
        return
      }
      for (const child of children) {
        this._render_child(child, el)
      }
    }
  }

  private _render_children(): void {
    for (const [node, children] of entries(this.model.children)) {
      const computed_children = (() => {
        if (isString(children)) {
          const value = this.model.data.property(children).get_value() as string | (string | UIElement)[]
          if (isString(value)) {
            return [value]
          } else if (isArray(value)) {
            return value
          } else {
            unreachable()
          }
        } else {
          return children
        }
      })()
      this._render_node(node, computed_children)
    }
  }

  private _render_html(literal: string, state: any = {}): any {
    let htm = literal.replace(/[`]/g, "\\$&")
    let collected_callbacks = ""
    const methods: string[] = []
    for (const [el_name, callbacks] of entries(this.model.callbacks)) {
      for (const [cb, method] of callbacks) {
        let definition: string
        htm = htm.replaceAll(`\${${method}}`, `$--{${method}}`)
        if (method.startsWith("script(")) {
          const meth = (
            method
              .replace("('", "_").replace("')", "")
              .replace('("', "_").replace('")', "")
              .replace("-", "_")
          )
          const script_name = meth.replaceAll("script_", "")
          htm = htm.replaceAll(method, meth)
          definition = `
          const ${meth} = (event) => {
            view._state.event = event
            view.run_script("${script_name}")
            delete view._state.event
          }
          `
        } else {
          definition = `
          const ${method} = (event) => {
            let elname = "${el_name}"
            if (RegExp("\{\{.*loop\.index.*\}\}").test(elname)) {
              const pattern = RegExp(elname.replace(/\{\{(.+?)\}\}/g, String.fromCharCode(92) + "d+"))
              for (const p of event.path) {
                if (pattern.exec(p.id) != null) {
                  elname = p.id.split("-").slice(null, -1).join("-")
                  break
                }
              }
            }
            view._send_event(elname, "${cb}", event)
          }
          `
        }
        if (methods.indexOf(method) > -1) {
          continue
        }
        methods.push(method)
        collected_callbacks += definition
      }
    }
    htm = htm
      .replaceAll("${model.", "$-{model.")
      .replaceAll("${", "${data.")
      .replaceAll("$-{model.", "${model.")
      .replaceAll("$--{", "${")
    return new Function("view, model, data, state, html, useCallback", `${collected_callbacks}return html\`${htm}\`;`)(
      this, this.model, this.model.data, state, html, useCallback,
    )
  }

  private _render_script(literal: string, id: string) {
    const scripts = []
    for (const elname of this.model.nodes) {
      const elvar = elname.replace("-", "_")
      if (!literal.includes(elvar)) {
        continue
      }
      const script = `
      let ${elvar} = view.shadow_el.getElementById('${elname}-${id}')
      if (${elvar} == null) {
        ${elvar} = document.getElementById('${elname}-${id}')
      }
      if (${elvar} == null) {
        console.warn("DOM node '${elname}' could not be found. Cannot execute callback.")
        return
      }
      `
      scripts.push(script)
    }
    const event = `
    let event = null
    if (state.event !== undefined) {
      event = state.event
    }
    `
    scripts.push(event)
    scripts.push(literal)
    return new Function("model, data, state, view, script, self", scripts.join("\n"))
  }

  private _remove_mutation_observers(): void {
    for (const observer of this._mutation_observers) {
      observer.disconnect()
    }
    this._mutation_observers = []
  }

  private _setup_mutation_observers(): void {
    const {id} = this.model.data
    for (const name of keys(this.model.attrs)) {
      const el = element_lookup(this.shadow_el, `${name}-${id}`)
      if (el == null) {
        console.warn(`DOM node '${name}-${id}' could not be found. Cannot set up MutationObserver.`)
        continue
      }
      const observer = new MutationObserver(() => {
        this._update_model(el, name)
      })
      observer.observe(el, {attributes: true})
      this._mutation_observers.push(observer)
    }
  }

  private _remove_event_listeners(): void {
    const {id} = this.model.data
    for (const [node, callbacks] of this._event_listeners) {
      const el = element_lookup(this.shadow_el, `${node}-${id}`)
      if (el == null) {
        continue
      }
      for (const [event_name, event_callback] of callbacks) {
        el.removeEventListener(event_name, event_callback)
      }
    }
    this._event_listeners.clear()
  }

  private _setup_event_listeners(): void {
    const {id} = this.model.data
    const attrs = dict(this.model.attrs)
    for (const [node, node_events] of entries(this.model.events)) {
      const el = element_lookup(this.shadow_el, `${node}-${id}`)
      if (el == null) {
        console.warn(`DOM node '${node}-${id}' could not be found. Cannot subscribe to DOM events.`)
        continue
      }
      for (const [event_name, event_doit] of entries(node_events)) {
        const event_callback = (event: Event) => {
          this._send_event(node, event_name, event)
          if (attrs.has(node) && event_doit) {
            this._update_model(el, node)
          }
        }
        el.addEventListener(event_name, event_callback)
        let callbacks = this._event_listeners.get(node)
        if (callbacks === undefined) {
          this._event_listeners.set(node, callbacks = new Map())
        }
        callbacks.set(event_name, event_callback)
      }
    }
  }

  private _update(property: string | null = null): void {
    if (property == null || this.html.includes(`\${${property}}`)) {
      const rendered = this._render_html(this.html)
      if (rendered == null) {
        return
      }
      try {
        this._changing = true
        render(rendered, this.container)
      } finally {
        this._changing = false
      }
    }
  }

  private _update_model(el: HTMLElement, name: string): void {
    if (this._changing) {
      return
    }
    const attr_infos = dict(this.model.attrs).get(name) ?? []
    const attrs: {[key: string]: unknown} = {}
    for (const attr_info of attr_infos) {
      const [attr, tokens, template] = attr_info
      let value: unknown = attr == "children" ? el.innerHTML : (el as any)[attr]
      if (tokens.length === 1 && (`{${tokens[0]}}` === template)) {
        attrs[tokens[0]] = value
      } else if (isString(value)) {
        value = extract_token(template, value, tokens)
        if (value == null) {
          console.warn(`Could not resolve parameters in ${name} element ${attr} attribute value ${value}.`)
        } else {
          for (const [param, param_val] of entries(value)) {
            if (param_val === undefined) {
              console.warn(`Could not resolve ${param} in ${name} element ${attr} attribute value ${value}.`)
            } else {
              attrs[param] = param_val
            }
          }
        }
      }
    }

    try {
      this._changing = true
      this.model.data.setv(serialize_attrs(attrs))
    } catch {
      console.error("Could not serialize", attrs)
    } finally {
      this._changing = false
    }
  }
}

export namespace ReactiveHTML {
  export type Attrs = p.AttrsOf<Props>

  export type Props = HTMLBox.Props & {
    attrs: p.Property<Dict<[string, string[], string][]>>
    callbacks: p.Property<Dict<[string, string][]>>
    children: p.Property<Dict<(UIElement | string)[] | string>>
    data: p.Property<Model>
    event_params: p.Property<string[]>
    events: p.Property<Dict<Dict<boolean>>>
    html: p.Property<string>
    looped: p.Property<string[]>
    nodes: p.Property<string[]>
    scripts: p.Property<Dict<string[]>>
  }
}

export interface ReactiveHTML extends ReactiveHTML.Attrs {}

export class ReactiveHTML extends HTMLBox {
  declare properties: ReactiveHTML.Props

  constructor(attrs?: Partial<ReactiveHTML.Attrs>) {
    super(attrs)
  }

  static override __module__ = "panel.models.reactive_html"

  static {
    this.prototype.default_view = ReactiveHTMLView
    this.define<ReactiveHTML.Props>(({Bool, Str, List, Dict, Tuple, Or, Ref}) => ({
      attrs:        [ Dict(List(Tuple(Str, List(Str), Str))), {} ],
      callbacks:    [ Dict(List(Tuple(Str, Str))), {} ],
      children:     [ Dict(Or(List(Or(Ref(UIElement), Str)), Str)), {} ],
      data:         [ Ref(Model) ],
      event_params: [ List(Str), [] ],
      events:       [ Dict(Dict(Bool)), {} ],
      html:         [ Str, "" ],
      looped:       [ List(Str), [] ],
      nodes:        [ List(Str), [] ],
      scripts:      [ Dict(List(Str)), {} ],
    }))
  }
}
